C++/Toolbox/Option/Option.h

```diff
@@ -0,0 +1,127 @@
+#ifndef _OPTION_H_
+#define _OPTION_H_
+
+#include <functional>
+#include <memory>
+#include <type_traits>
+
+namespace Toolbox {
+namespace Option {
+
+  /* The fundamental class describing an Option. Users only collect the Opt<T> as a result and do
+     do not instantiate this class directly. The static methods Some, None, and Option have 
+     standalone function aliases for convenience.
+
+     The underlying data structure is a std::unique_ptr. An Opt<T> is move only. 
+
+     The fundamental operations are forEach, map, and filter. `forEach` simply allows access to the
+     underlying data if it is Some. `map` and `filter` allow access and modification of the Opt<T>.
+  */ 
+  template<typename T>
+  class Opt
+  {
+  public:
+    // forward arguments to construct a Some<T>
+    template<typename... Args>
+    static Opt Some(Args... args)
+    {
+      return Option::Opt<T>(std::move(std::make_unique<T>(std::forward<Args>(args)...)));
+    }
+
+    // Create a None<T>
+    static Opt None()
+    {
+      return Option::Opt<T>();
+    }
+
+    // Create an Option from a pointer
+    // None<T> if is nullptr
+    // Else Some<T>
+    static Opt Option(T* t)
+    {
+      return Option::Opt<T>(std::move(std::unique_ptr<T>(t)));
+    }
+
+  private:
+    // Hidden constructors - one can't instantiate an Opt directly, only the class can
+    Opt(void) = default;
+    Opt(std::unique_ptr<T> p) : m_pData(std::move(p)) {};
```

Comment:
Curious, I noticed that on line 28, and 42 that we create the unique_ptr, move it, then in the constructor move it again.  Could we potentially instead, perfect forward the parameters to the constructor and have the constructor make the unique_ptr?

Reply:
Good idea, but there is an issue with it, and hopefully my elaboration will explain. If you see holes in my logic, feel free to point them out. 

Your idea is a good one. We could have a constructor that takes variadic paramter `Args&&... args`. But there is a problem: `static Opt Option(T* pT)` on line 40. When this method calls the constructor, how would it know which constructor to call? Would it call a constructor that takes a `T*` or the one with `Args&&...`? `std::unique_ptr` gets around this by having its constructor for the `T*` and `make_unique` for the `Args&&...`. The make function simply constructs a new instance and moves it into the `std::unique_ptr` constructor (which effectively takes a `T*`). 

I think what I am missing here though is some &&, so I should have written my constructor:
```
Opt(std::unique_ptr<T>&& p) : m_pData(std::move(p)) {};
```

Reply:
That makes perfect sense.  After reading this stack overflow explanation of passing unique_ptr's: [passing unique_ptr](https://stackoverflow.com/questions/8114276/how-do-i-pass-a-unique-ptr-argument-to-a-constructor-or-a-function), I believe your original implementation of passing my value is correct.  However, I believe that on line 28, std::make_unique returns an r-value and you do not need std::move?

Reply:
I think you're right, I will remove the unecessary std::move cast